### ISSUE TITLE
All Entities become "unavailable" after approximately 1 hour

### ISSUE THREAD

**Issue by mikedrews** [Tasshack/dreame-vacuum#721]
**Describe the bug**
Every Dreame integration entity becomes unavailable after approximately 1 hour.

**To Reproduce**
I wait for approximately 1 hour.  The integration must be reloaded in order for the entities to be refreshed or made available.

**Expected behavior**
Entities to be available.

**Screenshots**
![image](https://github.com/user-attachments/assets/f5ac2225-a221-4410-9b21-4b057190d623)

**Additional Information (please complete the following information)**
 - Model Name:  Dreame XL30 Ultra r9316k
 - Firmware Version [e.g. 1156]: 4.3.9_1726
 - Home Assistant Version: Core 2014.9.1 O/S 13.1
 - Configuration Type [With or without map support]: Dreame Account
 - Errors or warnings shown in the HA logs (if applicable):

The only errors I see by default are the following DNS errors.
[INFO] 127.0.0.1:55527 - 27810 "NS IN . udp 17 false 512" REFUSED qr,aa,rd 17 0.000061683s
[INFO] 127.0.0.1:36837 - 18322 "AAAA IN us.iot.dreame.tech. udp 47 true 2048" - - 0 30.001261635s
[ERROR] plugin/errors: 2 us.iot.dreame.tech. AAAA: dial tcp 1.1.1.1:853: i/o timeout
[INFO] 127.0.0.1:40178 - 5771 "A IN us.iot.dreame.tech. udp 47 true 2048" - - 0 30.001261698s
[ERROR] plugin/errors: 2 us.iot.dreame.tech. A: dial tcp 1.1.1.1:853: i/o timeout
[INFO] 172.30.32.1:40755 - 33633 "A IN us.iot.dreame.tech. udp 36 false 512" - - 0 8.002630949s
[ERROR] plugin/errors: 2 us.iot.dreame.tech. A: read udp 127.0.0.1:60776->127.0.0.1:5553: i/o timeout
[INFO] 172.30.32.1:40755 - 33810 "AAAA IN us.iot.dreame.tech. udp 36 false 512" - - 0 8.002917648s
[ERROR] plugin/errors: 2 us.iot.dreame.tech. AAAA: read udp 127.0.0.1:45978->127.0.0.1:5553: i/o timeout
[INFO] 127.0.0.1:51032 - 31159 "NS IN . udp 17 false 512" REFUSED qr,aa,rd 17 0.000081415s
[INFO] 127.0.0.1:54682 - 45305 "A IN us.iot.dreame.tech. udp 47 true 2048" - - 0 30.000378569s
[ERROR] plugin/errors: 2 us.iot.dreame.tech. A: dial tcp 1.1.1.1:853: i/o timeout
[INFO] 127.0.0.1:37568 - 45407 "AAAA IN us.iot.dreame.tech. udp 47 true 2048" - - 0 30.000378553s
[ERROR] plugin/errors: 2 us.iot.dreame.tech. AAAA: dial tcp 1.0.0.1:853: i/o timeout
[INFO] 127.0.0.1:32845 - 29181 "AAAA IN us.iot.dreame.tech. udp 47 true 2048" - - 0 30.000201887s
[ERROR] plugin/errors: 2 us.iot.dreame.tech. AAAA: dial tcp 1.1.1.1:853: i/o timeout
[INFO] 127.0.0.1:60006 - 16969 "A IN us.iot.dreame.tech. udp 47 true 2048" - - 0 30.000152291s
[ERROR] plugin/errors: 2 us.iot.dreame.tech. A: dial tcp 1.1.1.1:853: i/o timeout
[INFO] 127.0.0.1:57831 - 33500 "AAAA IN us.iot.dreame.tech. udp 47 true 2048" - - 0 30.000334348s
[ERROR] plugin/errors: 2 us.iot.dreame.tech. AAAA: dial tcp 1.0.0.1:853: i/o timeout
[INFO] 127.0.0.1:41216 - 8150 "A IN us.iot.dreame.tech. udp 47 true 2048" - - 0 30.000385717s
[ERROR] plugin/errors: 2 us.iot.dreame.tech. A: dial tcp 1.1.1.1:853: i/o timeout
[INFO] 127.0.0.1:53766 - 16539 "AAAA IN us.iot.dreame.tech. udp 47 true 2048" - - 0 19.431556798s
[INFO] 127.0.0.1:44538 - 4678 "A IN us.iot.dreame.tech. udp 47 true 2048" - - 0 19.431571704s
[ERROR] plugin/errors: 2 us.iot.dreame.tech. AAAA: dial tcp 1.0.0.1:853: connect: network is unreachable
[ERROR] plugin/errors: 2 us.iot.dreame.tech. A: dial tcp 1.0.0.1:853: connect: network is unreachable
[INFO] 127.0.0.1:50688 - 36599 "AAAA IN us.iot.dreame.tech. udp 47 true 2048" - - 0 30.000244126s
[ERROR] plugin/errors: 2 us.iot.dreame.tech. AAAA: dial tcp 1.1.1.1:853: i/o timeout
[INFO] 127.0.0.1:37031 - 17691 "A IN us.iot.dreame.tech. udp 47 true 2048" - - 0 30.000150386s
[ERROR] plugin/errors: 2 us.iot.dreame.tech. A: dial tcp 1.1.1.1:853: i/o timeout
[INFO] 127.0.0.1:51653 - 48859 "A IN us.iot.dreame.tech. udp 47 true 2048" - - 0 30.00098477s
[ERROR] plugin/errors: 2 us.iot.dreame.tech. A: dial tcp 1.0.0.1:853: i/o timeout
[INFO] 127.0.0.1:45282 - 9292 "AAAA IN us.iot.dreame.tech. udp 47 true 2048" - - 0 30.001699655s
[ERROR] plugin/errors: 2 us.iot.dreame.tech. AAAA: dial tcp 1.0.0.1:853: i/o timeout
[INFO] 127.0.0.1:37340 - 7260 "A IN us.iot.dreame.tech. udp 47 true 2048" - - 0 30.000363634s
[ERROR] plugin/errors: 2 us.iot.dreame.tech. A: dial tcp 1.1.1.1:853: i/o timeout
[INFO] 127.0.0.1:50953 - 60302 "AAAA IN us.iot.dreame.tech. udp 47 true 2048" - - 0 30.000362694s
[ERROR] plugin/errors: 2 us.iot.dreame.tech. AAAA: dial tcp 1.1.1.1:853: i/o timeout
[INFO] 127.0.0.1:38433 - 46687 "A IN us.iot.dreame.tech. udp 47 true 2048" - - 0 21.66733183s
[ERROR] plugin/errors: 2 us.iot.dreame.tech. A: read tcp 172.30.32.3:53422->1.1.1.1:853: i/o timeout
[INFO] 127.0.0.1:49765 - 29561 "A IN us.iot.dreame.tech. udp 47 true 2048" - - 0 30.000985441s
[ERROR] plugin/errors: 2 us.iot.dreame.tech. A: dial tcp 1.1.1.1:853: i/o timeout
[INFO] 127.0.0.1:51819 - 53524 "AAAA IN us.iot.dreame.tech. udp 47 true 2048" - - 0 30.000906347s
[ERROR] plugin/errors: 2 us.iot.dreame.tech. AAAA: dial tcp 1.1.1.1:853: i/o timeout
[INFO] 127.0.0.1:60278 - 43903 "AAAA IN 10000.mt.us.iot.dreame.tech. udp 56 true 2048" - - 0 30.000856317s
[ERROR] plugin/errors: 2 10000.mt.us.iot.dreame.tech. AAAA: dial tcp 1.0.0.1:853: i/o timeout
[INFO] 127.0.0.1:60835 - 62115 "A IN 10000.mt.us.iot.dreame.tech. udp 56 true 2048" - - 0 30.000856304s
[ERROR] plugin/errors: 2 10000.mt.us.iot.dreame.tech. A: dial tcp 1.0.0.1:853: i/o timeout
[INFO] 127.0.0.1:41375 - 28445 "A IN us.iot.dreame.tech. udp 47 true 2048" - - 0 30.000899684s
[ERROR] plugin/errors: 2 us.iot.dreame.tech. A: dial tcp 1.0.0.1:853: i/o timeout
[INFO] 127.0.0.1:37798 - 15745 "AAAA IN us.iot.dreame.tech. udp 47 true 2048" - - 0 30.000984844s
[ERROR] plugin/errors: 2 us.iot.dreame.tech. AAAA: dial tcp 1.0.0.1:853: i/o timeout
[INFO] 127.0.0.1:48107 - 2520 "AAAA IN us.iot.dreame.tech. udp 47 true 2048" - - 0 30.000370637s
[ERROR] plugin/errors: 2 us.iot.dreame.tech. AAAA: dial tcp 1.1.1.1:853: i/o timeout
[INFO] 127.0.0.1:38474 - 6718 "A IN us.iot.dreame.tech. udp 47 true 2048" - - 0 30.000370647s
[ERROR] plugin/errors: 2 us.iot.dreame.tech. A: dial tcp 1.1.1.1:853: i/o timeout
[INFO] 127.0.0.1:57905 - 1993 "A IN 10000.mt.us.iot.dreame.tech. udp 56 true 2048" - - 0 30.000942313s
[ERROR] plugin/errors: 2 10000.mt.us.iot.dreame.tech. A: dial tcp 1.1.1.1:853: i/o timeout
[INFO] 127.0.0.1:60115 - 55861 "AAAA IN 10000.mt.us.iot.dreame.tech. udp 56 true 2048" - - 0 30.000883206s
[ERROR] plugin/errors: 2 10000.mt.us.iot.dreame.tech. AAAA: dial tcp 1.1.1.1:853: i/o timeout
[INFO] 127.0.0.1:32920 - 39228 "AAAA IN us.iot.dreame.tech. udp 47 true 2048" - - 0 30.000412402s
[ERROR] plugin/errors: 2 us.iot.dreame.tech. AAAA: dial tcp 1.0.0.1:853: i/o timeout
[INFO] 127.0.0.1:50505 - 37890 "A IN us.iot.dreame.tech. udp 47 true 2048" - - 0 30.000343704s
[ERROR] plugin/errors: 2 us.iot.dreame.tech. A: dial tcp 1.1.1.1:853: i/o timeout
[INFO] 127.0.0.1:41550 - 52477 "AAAA IN 10000.mt.us.iot.dreame.tech. udp 56 true 2048" - - 0 30.000415663s
[ERROR] plugin/errors: 2 10000.mt.us.iot.dreame.tech. AAAA: dial tcp 1.1.1.1:853: i/o timeout
[INFO] 127.0.0.1:37411 - 58318 "A IN 10000.mt.us.iot.dreame.tech. udp 56 true 2048" - - 0 30.000329174s
[ERROR] plugin/errors: 2 10000.mt.us.iot.dreame.tech. A: dial tcp 1.1.1.1:853: i/o timeout
[INFO] 127.0.0.1:49673 - 15388 "AAAA IN 10000.mt.us.iot.dreame.tech. udp 56 true 2048" - - 0 30.00077641s
[ERROR] plugin/errors: 2 10000.mt.us.iot.dreame.tech. AAAA: dial tcp 1.1.1.1:853: i/o timeout
[INFO] 127.0.0.1:36497 - 32539 "A IN 10000.mt.us.iot.dreame.tech. udp 56 true 2048" - - 0 30.000812839s
[ERROR] plugin/errors: 2 10000.mt.us.iot.dreame.tech. A: dial tcp 1.1.1.1:853: i/o timeout
[INFO] 127.0.0.1:37815 - 14543 "AAAA IN 10000.mt.us.iot.dreame.tech. udp 56 true 2048" - - 0 30.000470814s
[ERROR] plugin/errors: 2 10000.mt.us.iot.dreame.tech. AAAA: dial tcp 1.0.0.1:853: i/o timeout
[INFO] 127.0.0.1:37474 - 36886 "A IN 10000.mt.us.iot.dreame.tech. udp 56 true 2048" - - 0 30.000611751s
[ERROR] plugin/errors: 2 10000.mt.us.iot.dreame.tech. A: dial tcp 1.1.1.1:853: i/o timeout

**Comment by Maxou44** [comment]
Same issue on my side 😅

**Comment by mikedrews** [comment]
> 10000.mt.us.iot.dreame.tech is the mqtt server adress and if it disconnects that means you dont have a stable intermet connection. Without mqtt connection integration cannot get updates therefore all entities become unavailable. Actual bug is that the entities will stuck unavailable even after mqtt is reconnected and I am aware of that issue for a long time.

I do not think it is because I do not have a stable internet connection.  I have many devices in my home that "talk" and I do not see any issues at the time the integration becomes unavailable.  Also after checking my router logs there is no service interruption at all during the same time period.  I suspect it may be the latency around the travel time to the mqtt server environment.  I have also run pings simultaneously to both 10000.mt.us.iot.dream.tech (47.254.33.232) and 1.1.1.1 and I see packets being dropped going to 47.254.33.232.

**Comment by ranathan14** [comment]
I have the same issue. After some time (a few hours), the app goes offline, and all entities become unavailable. It needs to be reloaded, and everything starts working again. With this behavior, it becomes unusable. (v2.0.0b16)

**Comment by zoic21** [comment]
Hello,

Same for me I need to reload integration to fix that (I think it's due to my network connection).

**Comment by mikedrews** [comment]
> @mikedrews you haven't responded my question on my previous comment...
> 
> `10000.mt.us.iot.dreame.tech. A: dial tcp 1.1.1.1:853: i/o timeout` How do you explain mqtt client disconnect and timeout on reconnect if you have a stable network?
> 
> Mqtt is a socket based connection with built in keep alive packet mechanism. If you loose your network connection even for fraction of a second that socket will be closed and trigger an mqtt disconnected event and causes the integration to think that the device has been disconnected.
> 
> When have I first implemented mqtt api of the dreame cloud I though it is a connection to the device over cloud but now I know that it is a relay api hosted on the cloud just to inform the app about changes. So even if your vacuum is disconnected from the internet integration still be able to connect to the mqtt server. Now I need to fully refactor availability detection of the device from integration but don't have to do that right now.
> 
> I always knew that there may be some issues with the dreamehome api implementation, that is the whole reason it is still supported only with the Beta so you should also consider that you are using an experimental project that has fully released yet.

Is it possible to prove that the client is closing the connection as a result of not getting a cloud response in the allotted time period?

**Comment by mikedrews** [comment]
> You should be more supportive than complainant if you want me to solve this problem. This whole project completely relies on my motivation and If I loose that you will loose control of your vacuum...

I am trying to be supportive.  I am not point fingers at your work but at the "dreame" cloud infrastructures capability to respond properly.  I just don't think that everyone having this issue should be thrown into a "you have a bad network" bucket.
I have spent many hours utilizing your work and creating a dashboard card below that my family can be happy with.  You have inspired me !
![image](https://github.com/user-attachments/assets/9050729d-db4e-4748-9514-9d54ce0d08a3)

**Comment by denfren** [comment]
The error says 
```
10000.mt.us.iot.dreame.tech. A: dial tcp 1.1.1.1:853: i/o timeout
```

I suspect that your issue is DNS related. 

1.1.1.1 is cloudflare, 853 is DoT (dns over tls) and the 'A' means A-Record (ipv4) for '10000.mt.us.iot.dreame.tech'

**Comment by zoic21** [comment]
> > Hello,
> > Same for me I need to reload integration to fix that (I think it's due to my network connection).
> > Temporary fix in automation :
> > ```
> > alias: Recharger Dreame si indisponible
> > description: ""
> > triggers:
> >   - entity_id:
> >       - sensor.l10s_ultra_status
> >     to: unavailable
> >     trigger: state
> >     for:
> >       hours: 0
> >       minutes: 4
> >       seconds: 0
> > conditions: []
> > actions:
> >   - action: homeassistant.reload_config_entry
> >     data: {}
> >     target:
> >       device_id: b84b28646457bc6761e0edd0af560be3
> > mode: single
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It's reload dreame device if status is unvailable for 4 min.
> 
> Can you share history of this automation like how many times it triggered and with how much intervals between so I can get more idea about the issue?

Hello,
Since I create this automation never... I got the issue 2 times at 7 o'clock, I create automation and for now problem never occur again....

**Comment by firstcolle** [comment]
I have the same issue with x40 ultra.
The only solution I found was to create an automation that executes the action dreame_vacuum.vacuum_request_map every hour.
Till now i never had a disconnectin again

**Comment by Cheerpipe** [comment]
Out of curiosity, ¿what region do you use Tasshack?

**Comment by mikedrews** [comment]
I am still have the same issue with v2.0.0b17.

**Comment by sim4dim** [comment]
Root cause of the recurring "device offline" / "Unable to discover the device over cloud": a TCP retransmit/SACK deadlock in the robot's cloud daemon (packet capture)
 
  TL;DR: On a Dreame L40 Ultra I captured the robot's cloud traffic at the gateway. The robot's persistent TCP connection to the Dreame cloud periodically enters a retransmit/SACK deadlock: it gets stuck endlessly re-sending a segment the server has already
  SACK-acknowledged (instead of the segment that's actually missing), never recovers, silently abandons the connection after ~100 s (no RST/FIN), and opens a fresh one. Each abandonment is the "offline" blip / Unable to discover the device over cloud. The WiFi
  link and path to the cloud are healthy (~80 ms RTT, −34 dBm). This is firmware-side, in the robot's cloud daemon — not a network or integration problem.

  Symptom — App shows "offline" intermittently and self-recovers; robot keeps working. In HA, entities go unavailable with Update Failed: Unable to discover the device over cloud; cold setup bounces setup_retry/setup_error until it catches a window.
  
  Environment (network ruled out) — dreame.vacuum.r2492b, fw 4.3.9_1752, Dreamehome. WiFi −34 dBm, association rock-solid (connected-time climbs 5+ min, zero deauths). Gateway: no blocks, no IPS, DNS/NTP fine. Path to cloud ~79–88 ms RTT.
  
  Findings
  1. Resolves 10000.mt.us.iot.dreame.tech → Aliyun us-west-1 (47.251.39.79/47.254.33.232) on TCP 19973 — healthy US endpoint.
  2. New SYN every 60–120 s (hundreds/day); old conn not closed first.
  3. The bug — representative trace:
  robot  -> seq 2071:2411 [P.]
  server -> ack 2411                    # server has it
  robot  -> seq 2071:2411 (retx)        # re-sends already-acked bytes
  server -> [sack 2071:2411]            # "have that block, waiting for something BELOW"
  robot  -> seq 2071:2411 (retx, exp backoff +1s,+2s,+5s,+9s...)
  ...                                    # never sends the actual gap; never recovers
  (~100s) robot -> SYN to the OTHER NLB IP   # abandons silently, reconnects
  3. Server keeps SACKing the same block; robot keeps retransmitting the already-acknowledged segment instead of the missing one. No RST/FIN — it just times out after ~100 s. A short probe connection (SYN→…→immediate FIN, zero data) precedes the NLB-IP switch.
  4. No clean teardown → abandoned ESTABLISHED states pile up (one device hoarded ~5,500 on my firewall until I added a per-source cap).

  Why it surfaces — During the deadlock window, cloud get_properties returns nothing → Unable to discover the device over cloud / entities unavailable; the app sees abandon→reconnect as offline→online.

  Scope — Matches the broad "device offline" reports across the Ultra line on the 4.3.9 firmware family (hourly unavailable + TCP read timeouts, "map blurry after reconnect", "unable to discover over cloud" on X40/XL30). Reproduced next to the router on
  unrelated networks → firmware-side, not WiFi.

  Caveats — single unit / few cycles; byte offsets illustrative, the pattern was consistent; TLS contents not inspected.
  
  Ask — Captures from other models/firmware would confirm scope. Looks like a SACK-recovery bug in the cloud daemon's TCP stack (retransmits the SACKed block instead of the gap, gives up after ~100 s) — fixable in firmware.
